cdc/generation.cc

```diff
@@ -154,137 +158,69 @@ topology_description generate_topology_description(
-
-        cdc_log.warn("Generation of CDC streams failed to create streams for some (vnode, shard) pair."
-                     " This can lead to worse performance.");
-
-        stream_id some_stream;
-        size_t idx = 0;
-        for (; idx < entries.size(); ++idx) {
-            for (auto s: entries[idx].streams) {
-                if (s.is_set()) {
-                    some_stream = s;
-                    break;
-                }
-            }
-            if (some_stream.is_set()) {
-                break;
-            }
+        t = dht::token::from_int64(std::numeric_limits<int64_t>::min());
+        if (sharder.shard_of(t) != shard_idx) {
+            t = sharder.token_for_next_shard(t, shard_idx);
         }
+        if (t > end) {
+            // This means a shard does not own any tokens in this vnode
+            // and it's ok to pick any token so we pick end of vnode.
+            return end;
+        }
+        return t;
+    });
 
-        assert(idx != entries.size() && some_stream.is_set());
-
-        // Iterate over all ranges in the clockwise direction, starting with the one we found a stream for.
-        for (size_t off = 0; off < entries.size(); ++off) {
-            auto& ss = entries[(idx + off) % entries.size()].streams;
-
-            int last_set_stream_idx = ss.size() - 1;
-            while (last_set_stream_idx > -1 && !ss[last_set_stream_idx].is_set()) {
-                --last_set_stream_idx;
-            }
-
-            if (last_set_stream_idx == -1) {
-                cdc_log.warn(
-                        "CDC wasn't able to generate any stream for vnode ({}, {}]. We'll use another vnode's streams"
-                        " instead. This might lead to inconsistencies.",
-                        tokens[(idx + off + entries.size() - 1) % entries.size()], tokens[(idx + off) % entries.size()]);
-
-                ss[0] = some_stream;
-                last_set_stream_idx = 0;
-            }
-
-            some_stream = ss[last_set_stream_idx];
-
-            // Replace 'unset' stream ids with indexes below last_set_stream_idx
-            for (int s_idx = last_set_stream_idx - 1; s_idx > -1; --s_idx) {
-                if (ss[s_idx].is_set()) {
-                    some_stream = ss[s_idx];
-                } else {
-                    ss[s_idx] = some_stream;
-                }
+    // Then we cover all vnodes that don't wrap around
+    for (size_t i = 1; i < tokens.size(); ++i) {
+        auto start = tokens[i - 1];
+        auto end = tokens[i];
+        setup_vnode(entries[i], start, end, [&] (const dht::sharder& sharder, size_t shard_idx) {
+            // It is guaranteed that start is not MAX_INT64 because end is still greater
+            auto t = dht::token::from_int64(dht::token::to_int64(start) + 1);
+            if (sharder.shard_of(t) != shard_idx) {
+                t = sharder.token_for_next_shard(start, shard_idx);
```

Comment:
this should also probably be part of `first_token_for_shard` (or however you name it)